File: packages/app-game-domain/src/app-game-notification-local-outbox-bridge.ts
import { type Infer, Schema, withParser } from '@ocentra-parent/schema-domain/effect';
import {
  AppGameNotificationDeliveryClaimState,
  AppGameNotificationIntentSchema,
  AppGameNotificationIntentStatus,
  type AppGameNotificationIntent,
} from './app-game-notification-intent';
import {
  NotificationLocalOutboxRecordSchema,
  type NotificationLocalOutboxRecord,
} from './notification-local-outbox-adapter-proof';
import { FamilyReferenceSchema, type FamilyReference, type ParentActionReference } from '@ocentra-parent/family-domain/references';
import {
  ParentContractSchemaVersion,
  ParentContractSchemaVersionSchema,
  ParentTimestampSchema,
} from '@ocentra-parent/family-domain/reference-primitives';
import { V3NotificationRuleReasonCodeSchema } from './v3-notification-rule-provider-retry-contract';
const NonEmptyAppGameNotificationOutboxBridgeText = Schema.String.pipe(Schema.minLength(1));
export const AppGameNotificationLocalOutboxBridgeStatus = {
  Linked: 'linked-local-outbox-record',
  ManualRequired: 'manual-required',
  Unavailable: 'unavailable',
} as const;
export const AppGameNotificationLocalOutboxBridgeStatusSchema = withParser(
  Schema.Literal(...Object.values(AppGameNotificationLocalOutboxBridgeStatus))
);
export const AppGameNotificationLocalOutboxBridgeIdSchema = NonEmptyAppGameNotificationOutboxBridgeText.pipe(
  Schema.brand('AppGameNotificationLocalOutboxBridgeId')
);
export const AppGameNotificationLocalOutboxBridgeReferenceSchema = NonEmptyAppGameNotificationOutboxBridgeText.pipe(
  Schema.brand('AppGameNotificationLocalOutboxBridgeReference')
);
const AppGameNotificationLocalOutboxBridgeRowBaseSchema = Schema.Struct({
  bridgeRecordId: AppGameNotificationLocalOutboxBridgeReferenceSchema,
  status: AppGameNotificationLocalOutboxBridgeStatusSchema,
  intent: AppGameNotificationIntentSchema,
  outboxRecord: Schema.Union(NotificationLocalOutboxRecordSchema, Schema.Null),
  blockedReasonRefs: Schema.Array(AppGameNotificationLocalOutboxBridgeReferenceSchema),
});
export const AppGameNotificationLocalOutboxBridgeRowSchema = withParser(
  AppGameNotificationLocalOutboxBridgeRowBaseSchema.pipe(
    Schema.filter(
      (row) =>
        appGameNotificationBridgeRowIsHonest(row) ||
        'Expected app/game notification outbox bridge rows to link only local-outbox-eligible intents and keep manual/unavailable intents out of queued records'
    )
  )
);
const AppGameNotificationLocalOutboxBridgeReadModelBaseSchema = Schema.Struct({
  schemaVersion: ParentContractSchemaVersionSchema,
  bridgeId: AppGameNotificationLocalOutboxBridgeIdSchema,
  generatedAt: ParentTimestampSchema,
  family: FamilyReferenceSchema,
  outboxRootRef: AppGameNotificationLocalOutboxBridgeReferenceSchema,
  rows: Schema.Array(AppGameNotificationLocalOutboxBridgeRowSchema),
  linkedRecordCount: Schema.Number.pipe(Schema.int(), Schema.nonNegative()),
  manualRequiredCount: Schema.Number.pipe(Schema.int(), Schema.nonNegative()),
  unavailableCount: Schema.Number.pipe(Schema.int(), Schema.nonNegative()),
  providerDeliveryRuntimeClaimed: Schema.Literal(false),
  providerReceiptIngestionClaimed: Schema.Literal(false),
  schedulerRuntimeClaimed: Schema.Literal(false),
  cloudRoutingClaimed: Schema.Literal(false),
  parentNotificationUiClaimed: Schema.Literal(false),
  childDeliveryClaimed: Schema.Literal(false),
  adapterDispatchClaimed: Schema.Literal(false),
});
export const AppGameNotificationLocalOutboxBridgeReadModelSchema = withParser(
  AppGameNotificationLocalOutboxBridgeReadModelBaseSchema.pipe(
    Schema.filter(
      (readModel) =>
        appGameNotificationBridgeReadModelCountsMatch(readModel) ||
        'Expected app/game notification outbox bridge counts to match linked manual-required and unavailable rows'
    )
  )
);
export type AppGameNotificationLocalOutboxBridgeStatus = Infer<typeof AppGameNotificationLocalOutboxBridgeStatusSchema>;
export type AppGameNotificationLocalOutboxBridgeRow = Infer<typeof AppGameNotificationLocalOutboxBridgeRowSchema>;
export type AppGameNotificationLocalOutboxBridgeReadModel = Infer<
  typeof AppGameNotificationLocalOutboxBridgeReadModelSchema
>;
export type AppGameNotificationLocalOutboxBridgeOptions = {
  readonly family: FamilyReference;
  readonly parentAction: ParentActionReference;
  readonly generatedAt: string;
  readonly bridgeId: string;
  readonly outboxRootRef: string;
  readonly outboxFileRef: string;
  readonly localDataPathRef: string;
};
export function buildAppGameNotificationLocalOutboxBridgeReadModel(
  options: AppGameNotificationLocalOutboxBridgeOptions,
  intents: ReadonlyArray<AppGameNotificationIntent>
): AppGameNotificationLocalOutboxBridgeReadModel {
  const rows = intents.map((intent) => appGameNotificationIntentToLocalOutboxBridgeRow(options, intent));
  return AppGameNotificationLocalOutboxBridgeReadModelSchema.parse({
    schemaVersion: ParentContractSchemaVersion.V0_6,
    bridgeId: options.bridgeId,
    generatedAt: options.generatedAt,
    family: options.family,
    outboxRootRef: options.outboxRootRef,
    rows,
    linkedRecordCount: countRows(rows, AppGameNotificationLocalOutboxBridgeStatus.Linked),
    manualRequiredCount: countRows(rows, AppGameNotificationLocalOutboxBridgeStatus.ManualRequired),
    unavailableCount: countRows(rows, AppGameNotificationLocalOutboxBridgeStatus.Unavailable),
    providerDeliveryRuntimeClaimed: false,
    providerReceiptIngestionClaimed: false,
    schedulerRuntimeClaimed: false,
    cloudRoutingClaimed: false,
    parentNotificationUiClaimed: false,
    childDeliveryClaimed: false,
    adapterDispatchClaimed: false,
  });
}
export function appGameNotificationIntentToLocalOutboxBridgeRow(
  options: AppGameNotificationLocalOutboxBridgeOptions,
  candidate: AppGameNotificationIntent
): AppGameNotificationLocalOutboxBridgeRow {
  const intent = AppGameNotificationIntentSchema.parse(candidate);
  const status = bridgeStatusForIntent(intent);

  return AppGameNotificationLocalOutboxBridgeRowSchema.parse({
    bridgeRecordId: `app-game-notification-outbox-bridge-${intent.notificationIntentId}`,
    status,
    intent,
    outboxRecord:
      status === AppGameNotificationLocalOutboxBridgeStatus.Linked ? localOutboxRecordForIntent(options, intent) : null,
    blockedReasonRefs:
      status === AppGameNotificationLocalOutboxBridgeStatus.Linked ? [] : intent.manualProofRequirements,
  });
}
export function serializeAppGameNotificationLocalOutboxJsonl(
  readModel: AppGameNotificationLocalOutboxBridgeReadModel
): string {
  return `${readModel.rows
    .flatMap((row) => (row.outboxRecord === null ? [] : [row.outboxRecord]))
    .map((record) => JSON.stringify(record))
    .join('\n')}\n`;
}
export function parseAppGameNotificationLocalOutboxJsonl(jsonl: string): NotificationLocalOutboxRecord[] {
  return jsonl
    .split('\n')
    .filter((line) => line.trim().length > 0)
    .map((line) => NotificationLocalOutboxRecordSchema.parse(JSON.parse(line)));
}
function bridgeStatusForIntent(intent: AppGameNotificationIntent): AppGameNotificationLocalOutboxBridgeStatus {
  if (intent.intentStatus === AppGameNotificationIntentStatus.LocalOutboxEligible) {
    return AppGameNotificationLocalOutboxBridgeStatus.Linked;
  }
  if (intent.intentStatus === AppGameNotificationIntentStatus.Unavailable) {
    return AppGameNotificationLocalOutboxBridgeStatus.Unavailable;
  }
  return AppGameNotificationLocalOutboxBridgeStatus.ManualRequired;
}
function localOutboxRecordForIntent(
  options: AppGameNotificationLocalOutboxBridgeOptions,
  intent: AppGameNotificationIntent
): NotificationLocalOutboxRecord {
  return NotificationLocalOutboxRecordSchema.parse({
    entryId: intent.localOutboxRecordRef,
    state: 'queued-local',
    envelope: {
      alertRef: `app-game-notification-alert-${intent.notificationIntentId}`,
      family: options.family,
      device: intent.device,
      parentAction: options.parentAction,
      severity: intent.priority,
      reasonCode: appGameNotificationReasonToProviderReason(intent.notificationReasonCode),
      providerChannel: intent.providerChannelPreference,
      evidenceRefs: intent.evidenceReferences,
      policyRefs: intent.policyRefs,
      auditRefs: intent.auditRefs,
      payloadTemplateRef: intent.notificationRuleRef,
      providerPayloadPreview:
        'alert id, family device scope, severity, reason code, evidence ref, policy ref, parent action link',
      sensitiveDetailMinimized: true,
      rawChildEvidenceIncluded: false,
      rawUrlOrTitleIncluded: false,
      rawMessageTextIncluded: false,
      screenshotOrReportIncluded: false,
    },
    outboxFileRef: options.outboxFileRef,
    localDataPathRef: options.localDataPathRef,
    deliveryClaimState: 'local-outbox-only',
    visibleAfterAt: null,
    retryAttemptCount: 0,
    quietHoursRef: null,
    retryPolicyRef: null,
    deadLetterRef: null,
    providerReceiptRef: null,
    manualProofRequirements: [],
    manualActionRequired: false,
    providerDeliveryAttempted: false,
    providerDeliveryObserved: false,
    providerReceiptIngested: false,
    providerCredentialsStored: false,
    cloudRoutingClaimed: false,
    parentNotificationUiClaimed: false,
    sensitiveProviderMetadataStored: false,
  });
}
function appGameNotificationReasonToProviderReason(reasonCode: string) {
  if (reasonCode === 'app-game-time-limit') {
    return V3NotificationRuleReasonCodeSchema.parse('policy-violation');
  }
  if (reasonCode === 'app-game-suspicious-unknown') {
    return V3NotificationRuleReasonCodeSchema.parse('suspicious-unknown');
  }
  return V3NotificationRuleReasonCodeSchema.parse('parent-request');
}
function appGameNotificationBridgeRowIsHonest(
  row: Infer<typeof AppGameNotificationLocalOutboxBridgeRowBaseSchema>
): boolean {
  if (row.status === AppGameNotificationLocalOutboxBridgeStatus.Linked) {
    return (
      row.intent.intentStatus === AppGameNotificationIntentStatus.LocalOutboxEligible &&
      row.intent.deliveryClaimState === AppGameNotificationDeliveryClaimState.LocalOutboxOnly &&
      row.outboxRecord !== null &&
      String(row.outboxRecord.entryId) === String(row.intent.localOutboxRecordRef) &&
      row.blockedReasonRefs.length === 0
    );
  }
  return row.outboxRecord === null && row.blockedReasonRefs.length > 0;
}
function appGameNotificationBridgeReadModelCountsMatch(
  readModel: Infer<typeof AppGameNotificationLocalOutboxBridgeReadModelBaseSchema>
): boolean {
  return (
    readModel.linkedRecordCount === countRows(readModel.rows, AppGameNotificationLocalOutboxBridgeStatus.Linked) &&
    readModel.manualRequiredCount ===
      countRows(readModel.rows, AppGameNotificationLocalOutboxBridgeStatus.ManualRequired) &&
    readModel.unavailableCount === countRows(readModel.rows, AppGameNotificationLocalOutboxBridgeStatus.Unavailable)
  );
}
function countRows(
  rows: ReadonlyArray<{ readonly status: AppGameNotificationLocalOutboxBridgeStatus }>,
  status: AppGameNotificationLocalOutboxBridgeStatus
): number {
  return rows.filter((row) => row.status === status).length;
}

File: crates/app-game-core/src/app_game_notification_local_outbox_bridge.rs
use std::io;

use ocentra_eventing::error::EventingError;
use ocentra_parent_agent_protocol::app_game::APP_GAME_SCHEMA_VERSION;
use ocentra_parent_agent_protocol::app_game_notification_readiness::AppGameNotificationReadinessRow;
use ocentra_parent_agent_protocol::schema_domain_mirrors::notification::NotificationLocalOutboxRecord;

use crate::app_game_child_ux_outbox_store::AppGameChildUxLocalOutboxStore;
use crate::app_game_child_ux_outbox_types::AppGameChildUxOutboxPersistResult;
use crate::app_game_notification_local_outbox_bridge_mapping::bridge_row;
use crate::app_game_notification_local_outbox_bridge_types::{
    AppGameNotificationLocalOutboxBridgeOptions, AppGameNotificationLocalOutboxBridgeReadModel,
    AppGameNotificationLocalOutboxBridgeStatus,
};
use crate::app_game_notification_local_outbox_bridge_validation::{
    validate_options, validate_source,
};

pub fn build_app_game_notification_local_outbox_bridge(
    options: AppGameNotificationLocalOutboxBridgeOptions,
    source_rows: Vec<AppGameNotificationReadinessRow>,
) -> Result<AppGameNotificationLocalOutboxBridgeReadModel, EventingError> {
    validate_options(&options)?;
    let rows = source_rows
        .into_iter()
        .map(|source| {
            validate_source(&source)?;
            bridge_row(&options, source)
        })
        .collect::<Result<Vec<_>, EventingError>>()?;
    let linked_record_count = count_rows(&rows, AppGameNotificationLocalOutboxBridgeStatus::Linked);
    let manual_required_count = count_rows(
        &rows,
        AppGameNotificationLocalOutboxBridgeStatus::ManualRequired,
    );
    let unavailable_count = count_rows(
        &rows,
        AppGameNotificationLocalOutboxBridgeStatus::Unavailable,
    );
    Ok(AppGameNotificationLocalOutboxBridgeReadModel {
        schema_version: APP_GAME_SCHEMA_VERSION,
        bridge_id: options.bridge_id,
        generated_at: options.generated_at,
        family: options.family,
        outbox_root_ref: options.outbox_root_ref,
        rows,
        linked_record_count,
        manual_required_count,
        unavailable_count,
        provider_delivery_runtime_claimed: false,
        provider_receipt_ingestion_claimed: false,
        scheduler_runtime_claimed: false,
        cloud_routing_claimed: false,
        parent_notification_ui_claimed: false,
        child_delivery_claimed: false,
        adapter_dispatch_claimed: false,
    })
}

pub fn persist_app_game_notification_local_outbox_bridge(
    store: &AppGameChildUxLocalOutboxStore,
    read_model: &AppGameNotificationLocalOutboxBridgeReadModel,
) -> io::Result<Vec<AppGameChildUxOutboxPersistResult>> {
    read_model
        .rows
        .iter()
        .filter_map(|row| row.outbox_record.clone())
        .map(|record| store.persist(record))
        .collect()
}

pub fn serialize_app_game_notification_local_outbox_jsonl(
    read_model: &AppGameNotificationLocalOutboxBridgeReadModel,
) -> Result<String, serde_json::Error> {
    let lines = read_model
        .rows
        .iter()
        .filter_map(|row| row.outbox_record.as_ref())
        .map(serde_json::to_string)
        .collect::<Result<Vec<_>, _>>()?;
    if lines.is_empty() {
        Ok(String::new())
    } else {
        Ok(format!("{}\n", lines.join("\n")))
    }
}

pub fn parse_app_game_notification_local_outbox_jsonl(
    jsonl: &str,
) -> Result<Vec<NotificationLocalOutboxRecord>, serde_json::Error> {
    jsonl
        .lines()
        .filter(|line| !line.trim().is_empty())
        .map(serde_json::from_str)
        .collect()
}

fn count_rows(
    rows: &[crate::app_game_notification_local_outbox_bridge_types::AppGameNotificationLocalOutboxBridgeRow],
    status: AppGameNotificationLocalOutboxBridgeStatus,
) -> u64 {
    rows.iter().filter(|row| row.status == status).count() as u64
}
